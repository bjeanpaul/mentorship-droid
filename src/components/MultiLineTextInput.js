import React, { Component } from 'react';

import {
  View,
  TextInput,
} from 'react-native';

import globalStyles from '../StyleSheet';


// TODO: Figure out how to animate the state changes
// TODO: Try and remove the magic numbers.
export default class MultiLineTextInput extends Component {

  constructor(props) {
    super(props);
    this.state = {
      height: 0,
    };
  }

  render() {
    return (
      <View style={globalStyles.textInputContainer}>
        <TextInput multiline enablesReturnKeyAutomatically
          {...this.props}
          underlineColorAndroid="transparent"
          onChange={(event) => {
            this.setState({
              height: event.nativeEvent.contentSize.height + 8,
            });
          }}
          style={[globalStyles.textInput, { height: Math.max(28, this.state.height) }]}
        />
      </View>
    );
  }

}
