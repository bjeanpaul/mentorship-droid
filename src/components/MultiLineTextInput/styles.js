import { StyleSheet } from 'react-native';


export default StyleSheet.create({
  charCount: {
    fontSize: 14,
    color: '#c0cbcc',
    textAlign: 'right',
    margin: 28,
    marginTop: -20,
    marginBottom: 0,
  },
});
