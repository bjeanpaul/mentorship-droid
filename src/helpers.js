import { find, isUndefined, partialRight as partial } from 'lodash';
import { getBaseURL } from 'src/configuration';
import { omitBy, isNull } from 'lodash';


export const omitNulls = d => omitBy(d, isNull);


export const trap = (type, fn) => v => v instanceof type
  ? Promise.resolve(v).then(fn)
  : Promise.reject(v);


export const switchError = cases => (e, ...xargs) => {
  const [type, fn] = find(cases, ([type, fn]) => e instanceof type) || [];
  return !isUndefined(fn)
    ? Promise.resolve().then(() => fn(e, ...xargs))
    : Promise.reject(e);
};


export const apiAction = ({
  method,
  request,
  success,
  failures,
}) => (...args) => (dispatch, { auth }) => Promise.resolve()
  .then(() => request(...args))
  .then(dispatch)
  .then(() => method(...args, auth))
  .then(partial(success, ...args), partial(switchError(failures), ...args))
  .then(dispatch);


export const staticAction = type => () => ({
  type
});


export const dataAction = type => data => ({
  type,
  payload: data
});


export const generateActionTypes = function generateActionNames(
  resource,
  operations = ['FETCH', 'CREATE', 'UPDATE', 'DELETE']
) {
  const actionTypes = {};
  const resourceUC = resource.toUpperCase();

  operations.forEach((operation) => {
    const operationUC = operation.toUpperCase();
    const request = `${resourceUC}_${operationUC}_REQUEST`;
    const success = `${resourceUC}_${operationUC}_SUCCESS`;
    const failure = `${resourceUC}_${operationUC}_FAILURE`;

    const operationLC = operation.toLowerCase();
    actionTypes[`${operationLC}Request`] = request;
    actionTypes[`${operationLC}Success`] = success;
    actionTypes[`${operationLC}Failure`] = failure;
  });
  return actionTypes;
};


export const filterActionTypes = function filterActionTypes(
  actionTypes,
  operation
) {
  const [request, success, failure] = [
    `${operation}Request`,
    `${operation}Success`,
    `${operation}Failure`,
  ];

  if (actionTypes[request] || actionTypes[success] || actionTypes[failure]) {
    return [
      actionTypes[request],
      actionTypes[success],
      actionTypes[failure],
    ];
  }
  return [];
};


// TODO: Delete
export const generateActionCreators = function generateActionCreators({
  resourcePath,
  actionTypes,
  requestOpts = {},
  normalizeJSON,
}) {
  const actionCreators = {};
  let url = `${getBaseURL()}/`;
  if (resourcePath) {
    url += `${resourcePath}/`;
  }

  const fetchActionTypes = filterActionTypes(actionTypes, 'fetch');
  if (fetchActionTypes.length === 3) {
    actionCreators.fetch = (onSuccess, onFailure) => ({
      type: '--generated fetch--',
      types: fetchActionTypes,
      url,
      requestOpts,
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }

  const createActionTypes = filterActionTypes(actionTypes, 'create');
  if (createActionTypes.length === 3) {
    actionCreators.create = ({
      body,
      onSuccess,
      onFailure,
    }) => ({
      type: '--generated create--',
      types: createActionTypes,
      url,
      requestOpts: { ...requestOpts,
        method: 'POST',
        body: JSON.stringify(body),
      },
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }

  // id is required as an argument to update a record;
  // we have a scenario with profiless, where it's
  // profile/{id}/image,
  // which is awkward...
  const updateActionTypes = filterActionTypes(actionTypes, 'update');
  if (updateActionTypes.length === 3) {
    actionCreators.update = ({
      id,
      body,
      onSuccess,
      onFailure,
    }) => ({
      type: '--generated update--',
      types: updateActionTypes,
      url: `${url}${id}/`,
      requestOpts: { ...requestOpts,
        method: 'PUT',
        body: JSON.stringify(body),
      },
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }


  return actionCreators;
};
