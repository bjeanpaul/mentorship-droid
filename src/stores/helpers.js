import moment from 'moment';
import { values, isUndefined, sortBy, filter, first } from 'lodash';


export const getAuthUserProfile = ({
  auth: { profileId },
  entities: { profiles },
}) => !isUndefined(profileId)
  ? profiles[profileId]
  : null;


export const getContext = state => {
  const { auth: { auth } } = state;

  return {
    auth,
    profile: getAuthUserProfile(state),
  };
};


// TODO sort?
export const getCategories = ({ entities: { categories } }) => values(categories);


// TODO sort?
export const getCategoryActivities = ({
  entities: { activities },
}, targetCategoryId) => (
  values(activities)
    .filter(({ category }) => category === targetCategoryId));


export const getCategory = ({ entities: { categories } }, id) => categories[id];


export const getActivity = ({ entities: { activities } }, id) => activities[id];


export const getScheduledCalls = ({ entities: { scheduledCalls } }) => values(scheduledCalls);


export const getScheduledCall = ({ entities: { scheduledCalls } }, id) => scheduledCalls[id];


export const getScheduledCallActivity = (state, id) => {
  const scheduledCall = getScheduledCall(state, id);
  return scheduledCall && scheduledCall.activity && getActivity(state, scheduledCall.activity);
};


export const getEvents = ({ entities: { events } }) => values(events);


export const getActivityCallNotes = ({
  entities: { callNotes },
}, activityId) => (
  filter(callNotes, { callActivity: activityId }));


export const getCall = ({ entities: { calls } }, id) => calls[id];


export const getCallNote = ({ entities: { callNotes } }, id) => callNotes[id];


export const getNextScheduledCall = (state, predicate = {}, time = Date.now()) => {
  let res = sortBy(getScheduledCalls(state), ({ callTime }) => callTime);
  res = res.filter(({ callTime }) => moment(callTime).isSameOrAfter(time));
  res = filter(res, predicate);
  return first(res);
};
