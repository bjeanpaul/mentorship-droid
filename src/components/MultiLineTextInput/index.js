import React, { PropTypes } from 'react';
import { View } from 'react-native';
import { TextInput, Text } from 'src/components';
import styles from './styles';


export default class MultiLineTextInput extends React.Component {
  constructor() {
    super();
    this.state = {
      height: 0,
      charCount: 0,
    };
  }

  render() {
    let charCount;
    if (this.props.maxLength) {
      charCount = (
        <Text style={styles.charCount}>
          {this.state.charCount}/{this.props.maxLength}
        </Text>);
    }

    return (
      <View>
        <TextInput
          {...this.props}
          multiline
          onChange={(event) => this.setState({
            height: event.nativeEvent.contentSize.height,
          })}
          onChangeText={(text) => {
            this.setState({
              charCount: text.length,
            });
            if (this.props.onChangeText) {
              this.props.onChangeText(text);
            }
          }}
          style={{
            height: Math.max(32, this.state.height),
          }}
        />
        {charCount}
      </View>
    );
  }
}


MultiLineTextInput.propTypes = {
  onChangeText: PropTypes.func,
  maxLength: PropTypes.number,
};
