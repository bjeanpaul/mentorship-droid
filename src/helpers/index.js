import { getBaseURL } from 'src/configuration';
import { omitBy, isNull } from 'lodash';


export const conj = (...args) => Object.assign({}, ...args);


export const omitNulls = d => omitBy(d, isNull);


export const generateActionTypes = function generateActionNames(
  resource,
  operations = ['FETCH', 'CREATE', 'UPDATE', 'DELETE']
) {
  const actionTypes = {};
  const resourceUC = resource.toUpperCase();

  operations.forEach((operation) => {
    const operationUC = operation.toUpperCase();
    const request = `${resourceUC}_${operationUC}_REQUEST`;
    const success = `${resourceUC}_${operationUC}_SUCCESS`;
    const failure = `${resourceUC}_${operationUC}_FAILURE`;

    const operationLC = operation.toLowerCase();
    actionTypes[`${operationLC}Request`] = request;
    actionTypes[`${operationLC}Success`] = success;
    actionTypes[`${operationLC}Failure`] = failure;
  });
  return actionTypes;
};


export const filterActionTypes = function filterActionTypes(
  actionTypes,
  operation
) {
  const [request, success, failure] = [
    `${operation}Request`,
    `${operation}Success`,
    `${operation}Failure`,
  ];

  if (actionTypes[request] || actionTypes[success] || actionTypes[failure]) {
    return [
      actionTypes[request],
      actionTypes[success],
      actionTypes[failure],
    ];
  }
  return [];
};


// TODO: Delete
export const generateActionCreators = function generateActionCreators({
  resourcePath,
  actionTypes,
  requestOpts = {},
  normalizeJSON,
}) {
  const actionCreators = {};
  let url = `${getBaseURL()}/`;
  if (resourcePath) {
    url += `${resourcePath}/`;
  }

  const fetchActionTypes = filterActionTypes(actionTypes, 'fetch');
  if (fetchActionTypes.length === 3) {
    actionCreators.fetch = (onSuccess, onFailure) => ({
      type: '--generated fetch--',
      types: fetchActionTypes,
      url,
      requestOpts,
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }

  const createActionTypes = filterActionTypes(actionTypes, 'create');
  if (createActionTypes.length === 3) {
    actionCreators.create = ({
      body,
      onSuccess,
      onFailure,
    }) => ({
      type: '--generated create--',
      types: createActionTypes,
      url,
      requestOpts: { ...requestOpts,
        method: 'POST',
        body: JSON.stringify(body),
      },
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }

  // id is required as an argument to update a record;
  // we have a scenario with profiless, where it's
  // profile/{id}/image,
  // which is awkward...
  const updateActionTypes = filterActionTypes(actionTypes, 'update');
  if (updateActionTypes.length === 3) {
    actionCreators.update = ({
      id,
      body,
      onSuccess,
      onFailure,
    }) => ({
      type: '--generated update--',
      types: updateActionTypes,
      url: `${url}${id}/`,
      requestOpts: { ...requestOpts,
        method: 'PUT',
        body: JSON.stringify(body),
      },
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }


  return actionCreators;
};
