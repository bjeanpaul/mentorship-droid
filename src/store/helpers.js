import moment from 'moment';
import { values, isUndefined, sortBy, filter, first, orderBy } from 'lodash';

import { getCurrent } from 'src/navigationHelpers';
import { TYPES_TO_COLLECTIONS } from 'src/constants/events';


export const getAuthUserProfile = ({
  auth: { profileId },
  entities: { profiles },
}) => !isUndefined(profileId)
  ? profiles[profileId]
  : null;


export const getContext = state => {
  const { auth: { auth } } = state;

  return {
    auth,
    profile: getAuthUserProfile(state),
  };
};


export const getCategories = ({ entities }) =>
  sortBy(values(entities.categories), 'ordinal');


export const getCategoryActivities = ({ entities }, targetCategoryId) => {
  const activities = values(entities.activities)
    .filter(({ category }) => category === targetCategoryId);

  return sortBy(activities, 'ordinal');
};


export const getCategory = ({ entities: { categories } }, id) =>
  categories[id];


export const getActivity = ({ entities: { activities } }, id) =>
  activities[id];


export const getScheduledCalls = ({ entities: { scheduledCalls } }) =>
  values(scheduledCalls);


export const getScheduledCall = ({ entities: { scheduledCalls } }, id) =>
  scheduledCalls[id];


export const getScheduledCallActivity = (state, id) => {
  const scheduledCall = getScheduledCall(state, id);
  return scheduledCall
      && scheduledCall.activity
      && getActivity(state, scheduledCall.activity);
};


export const getActivityCallNotes = ({
  entities: { callNotes },
}, activityId) =>
  values(callNotes)
    .filter(({ callActivityId }) => callActivityId === activityId);


export const getCallNotes = ({
  entities: { callNotes },
}) =>
  sortBy(callNotes, ({ callStartTime }) => +moment(callStartTime));


export const getCall = ({ entities: { calls } }, id) => calls[id];


export const getCallNote = ({ entities: { callNotes } }, id) => callNotes[id];


export const getNextScheduledCall = (state, predicate = {}, time = Date.now()) => {
  let res = sortBy(getScheduledCalls(state), ({ callTime }) => +moment(callTime));
  res = res.filter(({ callTime }) => moment(callTime).isSameOrAfter(time));
  res = filter(res, predicate);
  return first(res);
};


export const getScheduledCallsBetween = (state, start, end, time = Date.now()) => {
  // for some reason, moment durations are behaving inconsistently in tests,
  // thus the lack of moment usage here
  const startTime = +moment(time) - start;
  const endTime = +moment(time) + end;

  return getScheduledCalls(state)
    .filter(({ callTime }) => {
      const t = +moment(callTime);
      return startTime <= t && t <= endTime;
    });
};


const getEventObject = (state, event) => {
  const collection = state.entities[TYPES_TO_COLLECTIONS[event.eventType]];
  return collection && collection[event.objectId];
};


// TODO return the associated entitites to make the containers' jobs easier
export const getEvents = (state) =>
  values(state.entities.events)
    .filter(event => getEventObject(state, event));


export const getMessages = ({ entities }) => []
  .concat(values(entities.messages))
  .concat(values(entities.pendingMessages));


export const getMessage = ({ entities }, id) => entities.messages[id];


export const getActiveTopRoute = store => getCurrent(store.navigation.top);


export const getBlogPost = ({ entities }, id) => entities.blogPosts[id];


export const getBlogPosts = ({ entities }) =>
  orderBy(entities.blogPosts, [({ createdAt }) => +moment(createdAt)], ['desc']);
