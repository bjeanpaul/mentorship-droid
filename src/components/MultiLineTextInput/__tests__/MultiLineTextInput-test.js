import React from 'react';
import MultiLineTextInput from 'src/components/MultiLineTextInput';
import { uidEquals } from 'app/scripts/helpers';


describe('MultiLineTextInput', () => {
  const getCharCount = el => el
    .findWhere(uidEquals('charCount'))
    .children()
    .nodes.join('');

  it('should render', () => {
    expect(render(<MultiLineTextInput />)).toMatchSnapshot();
  });

  it('should show the number of characters used', () => {
    const el = shallow(<MultiLineTextInput value="21" maxLength={50} />);
    expect(getCharCount(el)).toEqual('2/50');

    el.findWhere(uidEquals('input'))
      .simulate('changeText', '2123');

    expect(getCharCount(el)).toEqual('4/50');
  });

  it('should call onChangeText() when text changes', () => {
    const onChangeText = jest.fn();
    const el = shallow(<MultiLineTextInput onChangeText={onChangeText} />);

    el.findWhere(uidEquals('input'))
      .simulate('changeText', '21');

    expect(onChangeText.mock.calls).toEqual([['21']]);
  });
});
