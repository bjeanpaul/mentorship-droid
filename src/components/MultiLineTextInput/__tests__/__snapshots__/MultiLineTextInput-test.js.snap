exports[`MultiLineTextInput should render 1`] = `
<View
  onLayout={undefined}>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
          "borderColor": "#dfe5e6",
          "marginBottom": 27,
          "marginLeft": 24,
          "marginRight": 24,
        },
        false,
      ]
    }>
    <TextInput
      multiline={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onContentSizeChange={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#c0cbcc"
      style={
        Array [
          Object {
            "color": "#003035",
            "fontFamily": "Brandon Text",
            "fontSize": 18,
            "marginBottom": 8,
            "padding": 0,
          },
          Object {
            "height": 0,
          },
        ]
      }
      uid="input"
      underlineColorAndroid="transparent" />
  </View>
</View>
`;
